### PR TITLE
Refactoring of the debugger

### PR DESCRIPTION
**Pull request by iFrostizz**
## Motivation

Initial issue: we cannot use breakpoints when launching the debugger through `forge debug`
This uncovered the fact that the debugger code is a bit confused.
For instance, when launching the debugger from `forge test`, we had to instantiate a `ScriptArgs` struct, which is the struct we were instantiating with `forge script --debug`.

## Solution

Create a new struct `DebuggerArgs` that any part of the code should comply with in order to launch the debugger.
Side effect: 
- fixes https://github.com/foundry-rs/foundry/issues/5510
- Adding breakpoints for fuzz (& invariant?) testing
- Reduces the debugger launching time because we don't have to run the test again (we do for fuzzing tho)

### CODE REVIEW

**Comment by iFrostizz** [comment]
- Adding breakpoints for fuzz & invariant testing

**Comment by mattsse** [comment]
heads up, we moved all crates to a crates folder,

sorry about the inconvenience

**Comment by iFrostizz** [comment]
Seems good, need to add back support for fuzz testing (optimized debugger for fuzzing it btw!!)
We may open a new PR of cleanups of the `Tui`. For instance instead of passing the source mappings and stuffs to the drawing functions, put them all in the `Tui` implementation.

**Comment by iFrostizz** [comment]
Could also bundle breakpoints and debug_arena together as they are passed around a lot together

File: crates/evm/src/fuzz/mod.rs [hunk 1]
```diff
@@ -101,76 +125,49 @@ impl<'a> FuzzedExecutor<'a> {
...
+
+                    if let Some(prev) = coverage.take() {
+                        // Safety: If `Option::or` evaluates to `Some`, then `call.coverage` must
+                        // necessarily also be `Some`
+                        coverage.replace(Some(prev.merge(case.coverage.unwrap())));
+                    } else {
+                        coverage.replace(case.coverage);
+                    }
+
+                    Ok(())
```
Comment (iFrostizz) [inline crates/evm/src/fuzz/mod.rs:148]: After this refactor (to be able to run only one fuzz if debug is on), it seems like the first rejection does not stop the test runner. Is that something related to `proptest` ?

**Comment by iFrostizz** [comment]
@Evalir @mattsse any idea about This issue about the fuzzer please ?
After this smol refactor for being able to `single_fuzz`, the fuzzer is super slow. Probably not returning on the first counter-example or something, but it might be specific to `proptest` and I'm not very familiar with this lib.

File: crates/evm/src/fuzz/mod.rs (hunk 1 above)
Comment (Evalir) [reply crates/evm/src/fuzz/mod.rs:148]: Right, I believe this might be something proptest related probably—it won't necessarily stop at the first rejection because it's looking for a counterexample and "shrinking" it—which means there might be something inefficient happening now. Maybe you could try flamegraphing this to see where it's taking so long? Also, here's the [proptest book](https://proptest-rs.github.io/proptest/intro.html)

Comment (iFrostizz) [reply crates/evm/src/fuzz/mod.rs:148]: @Evalir it was working as intended, I was just not using the `--release` build lmao

**Comment by iFrostizz** [comment]
Should be gtg! Source mapping on `cast run` are working also but don't seem to be very correct

**Comment by iFrostizz** [comment]
okay actually need to fix source mappings for called contracts (even locally)

**Comment by Evalir** [comment]
@iFrostizz amazing! was just wondering—this ended up being quite big haha, and touches a few sensitive things—was wondering if it's possible to split the PR into several smaller ones? it's ok if they all merge to an specific branch, it's just a lot to review

**Comment by iFrostizz** [comment]
Yes, good idea @Evalir Sorry for that it was not planned at all but appeared to be necessary. I'll break that into multiple PRs, how much do you wanna ideally ? +-500 lines ?

**Comment by Evalir** [comment]
yup around 500 lines is good! no worries—we're all guilty of big prs, this is great work!

**Comment by Evalir** [comment]
wondering @iFrostizz — is this closeable as this was split?

**Comment by iFrostizz** [comment]
Yep @Evalir , let's close!
